### PR TITLE
Separate diff decision from presentation

### PR DESCRIPTION
**Pull request by VenelinMartinov**
Adds a `DiffEqualDecisionOverride` to the shim layer which allows provider implementations to specify a diff decision instead of relying on the shim layer to do that.

Also adds a `DiffEqualDecisionOverride` to the PRC sdkv2 implementation which opentofu does.

Also adds a feature flag `EnableAccurateBridgePreview` which guards this feature.

will fix once rolled out: https://github.com/pulumi/pulumi-terraform-bridge/issues/2293
will fix once rolled out: https://github.com/pulumi/pulumi-terraform-bridge/issues/1501
will undo once rolled out:  https://github.com/pulumi/pulumi-terraform-bridge/pull/1502 as the underlying issue was fixed during the PRC work.
Stacked on https://github.com/pulumi/pulumi-terraform-bridge/pull/2380

### CODE REVIEW

**Comment by VenelinMartinov** [comment]
Kicked off downstream tests from [beec4df](https://github.com/pulumi/pulumi-terraform-bridge/pull/2379/commits/beec4df2d17bda727e5a93a36e6646dc19b5a80b)

File: pkg/tests/schema_pulumi_test.go [hunk 1]
```diff
@@ -2145,8 +2134,6 @@ Resources:
         [id=newid]
         [urn=urn:pulumi:test::test::prov:index/test:Test::mainRes]
       ~ listProps: [
-            [0]: "val1"
```
Comment (VenelinMartinov) [inline pkg/tests/schema_pulumi_test.go:2148]: Removing the collection from the detailed diff removes the unchanged items in it:
https://github.com/pulumi/pulumi-terraform-bridge/blob/cf59a2592ae7257786c441bf1a85bddad0246940/pkg/tfbridge/provider.go#L1159

Can be added back if we decide, will leave it away for now.

Comment (VenelinMartinov) [reply pkg/tests/schema_pulumi_test.go]: reverting from this PR as out of scope.

**Comment by VenelinMartinov** [comment]
Testing this in GCP and AWS:
https://github.com/pulumi/pulumi-gcp/pull/2362
https://github.com/pulumi/pulumi-aws/pull/4429

**Comment by VenelinMartinov** [comment]
GCP passes https://github.com/pulumi/pulumi-gcp/pull/2367#issuecomment-2333897203
Still waiting on AWS: https://github.com/pulumi/pulumi-aws/pull/4438

File: pkg/tfshim/shim.go [hunk 2]
```diff
@@ -48,6 +48,7 @@ type InstanceDiff interface {
 	ProposedState(res Resource, priorState InstanceState) (InstanceState, error)
 	Destroy() bool
 	RequiresNew() bool
+	DiffEqualDecisionOverride() *bool
```
Comment (t0yv0) [inline pkg/tfshim/shim.go:51]: Trying to understand this, how is this different from HasNoChanges()? Why would you need both in the model, can you give an example?

File: pkg/tfshim/sdk-v2/provider2.go [hunk 3]
```diff
@@ -262,13 +267,20 @@ func (p *planResourceChangeImpl) Diff(
 		return nil, err
 	}
 
+	//nolint:lll
+	// https://github.com/opentofu/opentofu/blob/864aa9d1d629090cfc4ddf9fdd344d34dee9793e/internal/tofu/node_resource_abstract_instance.go#L1024
+	unmarkedPrior, _ := st.UnmarkDeep()
```
Comment (t0yv0) [inline pkg/tfshim/sdk-v2/provider2.go:272]: contract.AssertNoError(err) on the errors?

**Comment by t0yv0** [comment]
I've added some nits on the code style but also a couple of higher level questions:

1. Are corner cases here where we introduced original hacks  already covered well by tests, can you point out which ones, I'd like to give it a read to double-check. 

2. Feels like a relatively big deal, originally we talked about rolling this under the "accurate previews" flag, but this PR suggests you're thinking maybe including it in a regular bridge update. Perhaps we have enough confidence to do that, just was making sure it's an informed decision.

3. Out of scope for this PR, but I wonder if you've noticed collectionDiffs feature; that seems to be still hanging around but I was hoping we can eventually remove this (maybe not yet in this PR), WDYT? It was about explaining these "mysterious diffs" which we computed incorrectly.  Perhaps we can file a reminder to clean it out in the Accurate Previews epic. 

```
	if !diff.HasNoChanges() {
		changes = pulumirpc.DiffResponse_DIFF_SOME
		// Perhaps collectionDiffs can shed some light and locate the changes to the end-user.
		for path, diff := range dd.collectionDiffs {
			detailedDiff[path] = diff
		}
	}

```

File: pkg/tfshim/shim.go (hunk 2 above)
Comment (VenelinMartinov) [reply pkg/tfshim/shim.go:53]: Yeah, `DiffEqualDecisionOverride` is like a feature flag - shim implementations which return a non-null value signify that this decision needs to be respected.

`HasNoChanges` seems to only be used in tests - we could rename `DiffEqualDecisionOverride` to something else and use `HasNoChanges` for the result. Or we could clean up `HasNoChanges`

File: pkg/tfshim/sdk-v2/provider2.go (hunk 3 above)
Comment (VenelinMartinov) [reply pkg/tfshim/sdk-v2/provider2.go:280]: The second value is not errors but the list of marks - we don't need that

**Comment by VenelinMartinov** [comment]
Hm, looking at this again, perhaps we can not merge this safely: This changes the Diff/NoDiff decision from being derived from the detailedDiff.

I thought this only affected the workaround added in https://github.com/pulumi/pulumi-terraform-bridge/blob/4954882afe708bad8c3f0d1859003dc5b05bb1f7/pkg/tfbridge/provider.go#L1149 but it does have bigger implications - might be best to hold off here and merge as part of the bigger changes.

**Comment by VenelinMartinov** [comment]
On your questions:

>Are corner cases here where we introduced original hacks already covered well by tests, can you point out which ones, I'd like to give it a read to double-check.
Did some delving into the git history and here is my understanding:
 - https://github.com/pulumi/pulumi-terraform-bridge/pull/1502 was added due to issues in GCP and AWS 
 - The GCP issue is related to labels and covered by the labels test in https://github.com/pulumi/pulumi-gcp/pull/1310/files#diff-2e4fffe32720b3d310b4336cfdde6519ab1bf6d9e0d82b418299817fdf42a29fR191
 - The AWS issue is related to empty tags: https://github.com/pulumi/pulumi-aws/pull/2944 - also covered by tests in AWS
 - Both the providers' test passed with this change enabled. That makes me believe that this issue is fixed and the workaround is no longer needed.

>Feels like a relatively big deal, originally we talked about rolling this under the "accurate previews" flag, but this PR suggests you're thinking maybe including it in a regular bridge update. Perhaps we have enough confidence to do that, just was making sure it's an informed decision.

Yeah, upon revisiting, this might not be safe to merge without a feature flag, will hold off.

>Out of scope for this PR, but I wonder if you've noticed collectionDiffs feature; that seems to be still hanging around but I was hoping we can eventually remove this (maybe not yet in this PR), WDYT? It was about explaining these "mysterious diffs" which we computed incorrectly. Perhaps we can file a reminder to clean it out in the Accurate Previews epic.

Yeah, noticed it but it is out of scope for this PR - here I only wanted to separate the Diff/NoDiff decision from the detailed diff - the detailed diff issues should be addressed as part of https://github.com/pulumi/pulumi-terraform-bridge/issues/2294

File: pkg/tfshim/shim.go (hunk 2 above)
Comment (iwahbe) [reply pkg/tfshim/shim.go:55]: > `HasNoChanges` seems to only be used in tests

I don't think this is true. I have seen diff status corrections from `HasNoChanges` in the wild.

Comment (VenelinMartinov) [reply pkg/tfshim/shim.go:55]: My bad, I deleted its only usage and then could not find it being used outside of tests...

Comment (VenelinMartinov) [reply pkg/tfshim/shim.go:55]: Do you have concrete examples of this @iwahbe? The only ones I could find which were in GCP labels and AWS tags.

**Comment by VenelinMartinov** [comment]
I've added a feature flag for this, along with an env var for it.

I'll set up CI to run with the feature flag so that we do not repeat mistakes from PRC

File: pkg/tfbridge/info/info.go [hunk 4]
```diff
@@ -159,6 +159,9 @@ type Provider struct {
 	// EnableZeroDefaultSchemaVersion makes the provider default
 	// to version 0 when no version is specified in the state of a resource.
 	EnableZeroDefaultSchemaVersion bool
+	// EnableAccurateBridgePreview makes the bridge use an experimental feature
+	// to generate more accurate diffs and previews for resources
+	EnableAccurateBridgePreview bool
```
[the remark this answers is not in the document]
Comment (VenelinMartinov) [reply pkg/tfbridge/info/info.go:164]: The plan for the flag is to guard both this and the soon to be added detailed diff v2 - I think `EnableAccurateBridgePreview` is a decent name for the package of features - LMK if you have something better

File: pkg/tfshim/shim.go [hunk 5]
```diff
@@ -50,6 +50,9 @@ type InstanceDiff interface {
 	ProposedState(res Resource, priorState InstanceState) (InstanceState, error)
 	Destroy() bool
 	RequiresNew() bool
+
+	// DiffEqualDecisionOverride can return a non-null value to override the default decision of if the diff is equal.
+	DiffEqualDecisionOverride() *bool
```
[the remark this answers is not in the document]
Comment (VenelinMartinov) [reply pkg/tfshim/shim.go]: Yeah, good suggestion

File: pkg/tfshim/sdk-v2/provider2.go [hunk 6]
```diff
@@ -269,13 +274,22 @@ func (p *planResourceChangeImpl) Diff(
 		TfToken:        t,
 		PlanState:      plan.PlannedState,
 	})
+
+	//nolint:lll
+	// https://github.com/opentofu/opentofu/blob/864aa9d1d629090cfc4ddf9fdd344d34dee9793e/internal/tofu/node_resource_abstract_instance.go#L1024
+	unmarkedPrior, _ := st.UnmarkDeep()
+	unmarkedPlan, _ := plannedState.UnmarkDeep()
+	eqV := unmarkedPrior.Equals(unmarkedPlan)
+	eq := eqV.IsKnown() && eqV.True()
```
[the remark this answers is not in the document]
Comment (VenelinMartinov) [reply pkg/tfshim/sdk-v2/provider2.go:286]: Added a comment for this.

File: pkg/tfshim/shim.go (hunk 2 above)
Comment (t0yv0) [reply pkg/tfshim/shim.go]:
```
pkg/tfshim/shim.go
49:	HasNoChanges() bool
pkg/tfbridge/provider.go
1158:	if !diff.HasNoChanges() {
pkg/tfshim/tfplugin5/provider_test.go
1156:			if diff.HasNoChanges() {

```

It's used. IDE to the rescue.

**Comment by t0yv0** [comment]
> I'll set up CI to run with the feature flag so that we do not repeat mistakes from the PRC rollout

Excellent, I was just jotting down a ticket to make tests go in a matrix fashion against the flag state space. Thank you.

**Review by t0yv0 — APPROVED** [review]
Thank you! Very important piece for us.

File: pkg/tfshim/shim.go (hunk 2 above)
Comment (VenelinMartinov) [reply pkg/tfshim/shim.go]: They are implemented differently - `HasNoChanges` used the TF Diff which we are trying to get rid of. I think for now we have to keep both around
